File: trees/BinarySearchTree.java
import java.util.Scanner;

public class BinarySearchTree{
    public Node root;

    static class Node{
        long element;
        Node left;
        Node right;

        Node(){
            left = null;
            right = null;
        }
    }

    public BinarySearchTree(long[] values){
        root = new Node();
        root.element = values[0];

        for(int i = 1;i<values.length;i++){
            Node current = root;
            while(true){
                if(current.element>values[i]){
                    if(current.left==null){
                        Node newNode = new Node();
                        current.left = newNode;
                        newNode.element = values[i];
                        break;
                    }else{
                        current = current.left;
                        continue;
                    }
                }else{
                    if(current.right==null){
                        Node newNode = new Node();
                        current.right=newNode;
                        newNode.element = values[i];
                        break;
                    }else{
                        current = current.right;
                        continue;
                    }
                }
            }
        }
    }

    public Node find(long element, Node n){
        if(element == n.element){
            return n;
        }
        if(element>n.element){
            return find(element, n.right);
        }else{
            return find(element, n.left);
        }
    }

    public void print(long element){
        Node n = find(element, this.root);
        print(n);
    }

    public void print(Node n){
        System.out.println(n.element);
        if(n.left != null){
            print(n.left);
        }
        if(n.right != null){
            print(n.right);
        }
    }

    public static void main(String[] args){
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        long A[] = new long[n];
        for (int i=0;i<n;i++){
            A[i] = in.nextLong();
        }
        BinarySearchTree bst = new BinarySearchTree(A);
        long t = in.nextLong();
        bst.print(t);
    }
}

File: trees/BinarySearchTree.js
process.stdin.resume();
process.stdin.setEncoding("utf-8");
var stdin_input = "";

process.stdin.on("data", function (input) {
    stdin_input += input;
});

process.stdin.on("end", function () {
   main(stdin_input);
});

class Tree{
    constructor(startElement){
        this.start = new Node(startElement, this);
    }

    print(node){
        if(node==null){
            return;
        }
        this.print(node.left);
        console.log(node.element);
        this.print(node.right);
    }

    find(element, node){
        if(node.element == element){
            return node;
        }
        if(node.element < element){
            if(node.right == null){
                return node;
            }else{
                return this.find(element, node.right);
            }
        }else{
            if(node.left == null){
                return node;
            }else{
                return this.find(element, node.left);
            }
        }
    }

    add(element){
        let node = this.find(element, this.start);
        if(node.element<element){
            node.right = new Node(element, node);
        }else{
            node.left = new Node(element, node);
        }
    }
}

class Node{
    constructor(element, parent){
        this.element = element;
        this.left = null;
        this.right = null;
        this.parent = parent;
    }
}

function main(input) {
    input = input.trim();
    let n = Number(input[0]);
    let es = input.slice(1, input.length - 2);
    let tree = null;
    for(let i=0;i<es.length;i++){
        let temp = Number(es[i].trim());
        if(!tree){
            tree = new Tree(temp);
        }else{
            tree.add(temp);
        }
    }
    let toFind = Number(input[input.length-2].trim())
    let node = tree.find(
        toFind, tree.start
    );
    tree.print(
        node
    );
}